// @smartledger/bsv/lib/smart_contract/utxo_generator.js
/**
 * SmartContract.UTXOGenerator Class
 * =================================
 * 
 * Enhanced UTXO generation with real BSV private/public keys
 * for authentic smart contract testing and development.
 * 
 * Features:
 * - Generate real BSV keypairs for testing
 * - Create authentic transaction inputs/outputs
 * - Support multiple script types (P2PKH, P2SH, custom)
 * - Integrate with existing SmartUTXO system
 * - Enable local testing with real cryptography
 */

'use strict'

var bsv = require('../..')
var crypto = require('crypto')

/**
 * UTXOGenerator Class - Real BSV UTXO generation for testing
 * @param {Object} options - Configuration options
 */
function UTXOGenerator(options) {
  if (!(this instanceof UTXOGenerator)) {
    return new UTXOGenerator(options)
  }

  this.options = options || {}
  this.network = this.options.network || bsv.Networks.mainnet
  this.keyRing = {} // Store generated keys
  this.utxoPool = [] // Store generated UTXOs
  
  // Initialize with SmartUTXO integration
  if (typeof bsv.SmartUTXO !== 'undefined') {
    this.smartUTXO = new bsv.SmartUTXO(this.options)
  }
}

/**
 * Generate a new BSV keypair for testing
 * @param {string} label - Optional label for the keypair
 * @returns {Object} Keypair with privateKey, publicKey, address
 */
UTXOGenerator.prototype.generateKeypair = function(label) {
  label = label || 'key_' + Date.now()
  
  var privateKey = bsv.PrivateKey.fromRandom(this.network)
  var publicKey = privateKey.toPublicKey()
  var address = privateKey.toAddress(this.network)
  
  var keypair = {
    label: label,
    privateKey: privateKey,
    publicKey: publicKey,
    address: address,
    wif: privateKey.toWIF(),
    addressString: address.toString()
  }
  
  // Store for later use
  this.keyRing[label] = keypair
  
  return keypair
}

/**
 * Create real UTXOs with authentic BSV transactions
 * @param {Object} config - UTXO configuration
 * @returns {Array} Array of real UTXOs
 */
UTXOGenerator.prototype.createRealUTXOs = function(config) {
  config = config || {}
  
  var utxoCount = config.count || 3
  var satoshisPerUTXO = config.satoshis || 100000
  var scriptType = config.scriptType || 'P2PKH'
  var keypair = config.keypair || this.generateKeypair('utxo_owner')
  
  var utxos = []
  
  for (var i = 0; i < utxoCount; i++) {
    var utxo = this._createSingleUTXO({
      keypair: keypair,
      satoshis: satoshisPerUTXO,
      scriptType: scriptType,
      vout: i
    })
    
    utxos.push(utxo)
    this.utxoPool.push(utxo)
    
    // Add to SmartUTXO system if available
    if (this.smartUTXO) {
      this.smartUTXO.addUTXO(utxo)
    }
  }
  
  return utxos
}

/**
 * Create a single authentic UTXO
 * @private
 */
UTXOGenerator.prototype._createSingleUTXO = function(config) {
  // Generate realistic transaction ID
  var txid = crypto.randomBytes(32).toString('hex')
  
  // Create appropriate script based on type
  var script
  var scriptHex
  
  switch (config.scriptType) {
    case 'P2PKH':
      script = bsv.Script.buildPublicKeyHashOut(config.keypair.address)
      scriptHex = script.toHex()
      break
      
    case 'P2SH':
      // Create a simple multisig for P2SH example
      var redeemScript = bsv.Script.buildMultisigOut([config.keypair.publicKey], 1)
      script = bsv.Script.buildScriptHashOut(redeemScript)
      scriptHex = script.toHex()
      break
      
    case 'CUSTOM':
      // Allow custom script injection
      script = config.customScript || bsv.Script.buildPublicKeyHashOut(config.keypair.address)
      scriptHex = script.toHex()
      break
      
    default:
      script = bsv.Script.buildPublicKeyHashOut(config.keypair.address)
      scriptHex = script.toHex()
  }
  
  return {
    txid: txid,
    vout: config.vout,
    address: config.keypair.addressString,
    script: scriptHex,
    satoshis: config.satoshis,
    keypair: config.keypair,
    scriptType: config.scriptType,
    scriptObj: script,
    created: new Date().toISOString()
  }
}

/**
 * Create a realistic transaction using generated UTXOs
 * @param {Object} config - Transaction configuration
 * @returns {Object} Transaction and signing details
 */
UTXOGenerator.prototype.createTestTransaction = function(config) {
  config = config || {}
  
  // Get UTXOs to spend
  var inputUTXOs = config.inputs || this.utxoPool.slice(0, 1)
  if (inputUTXOs.length === 0) {
    throw new Error('No UTXOs available for transaction. Call createRealUTXOs() first.')
  }
  
  // Calculate input total
  var inputTotal = inputUTXOs.reduce(function(sum, utxo) {
    return sum + utxo.satoshis
  }, 0)
  
  // Create transaction
  var transaction = new bsv.Transaction()
  
  // Add inputs
  inputUTXOs.forEach(function(utxo) {
    transaction.from({
      txId: utxo.txid,
      outputIndex: utxo.vout,
      address: utxo.address,
      script: utxo.script,
      satoshis: utxo.satoshis
    })
  })
  
  // Add outputs
  var outputAddress = config.outputAddress || inputUTXOs[0].keypair.addressString
  var outputAmount = config.outputAmount || (inputTotal - 10000) // Leave 10k sats for fee
  var fee = config.fee || 10000
  
  transaction.to(outputAddress, outputAmount)
  transaction.fee(fee)
  
  return {
    transaction: transaction,
    inputUTXOs: inputUTXOs,
    unsignedHex: transaction.toString(),
    inputTotal: inputTotal,
    outputAmount: outputAmount,
    fee: fee,
    
    // Signing helper
    sign: function() {
      inputUTXOs.forEach(function(utxo) {
        transaction.sign(utxo.keypair.privateKey)
      })
      return {
        signedTransaction: transaction,
        signedHex: transaction.toString(),
        txid: transaction.id
      }
    },
    
    // Preimage generation helper
    generatePreimage: function(inputIndex, sighashType) {
      sighashType = sighashType || bsv.crypto.Signature.SIGHASH_ALL | bsv.crypto.Signature.SIGHASH_FORKID
      var utxo = inputUTXOs[inputIndex]
      
      return bsv.Transaction.sighash.sighashPreimage(
        transaction,
        sighashType,
        inputIndex,
        bsv.Script.fromHex(utxo.script),
        new bsv.crypto.BN(utxo.satoshis)
      )
    }
  }
}

/**
 * Create covenant-ready UTXOs with preimage generation
 * @param {Object} config - Covenant configuration
 * @returns {Object} Covenant test setup
 */
UTXOGenerator.prototype.createCovenantTest = function(config) {
  config = config || {}
  
  // Generate keypair for covenant
  var covenantKeypair = this.generateKeypair('covenant_test')
  
  // Create UTXOs
  var utxos = this.createRealUTXOs({
    count: config.utxoCount || 2,
    satoshis: config.satoshis || 50000,
    keypair: covenantKeypair,
    scriptType: config.scriptType || 'P2PKH'
  })
  
  // Create test transaction
  var txConfig = {
    inputs: utxos.slice(0, 1), // Use first UTXO
    outputAmount: config.covenantAmount || 40000,
    fee: 10000
  }
  
  var testTx = this.createTestTransaction(txConfig)
  
  // Generate preimage for covenant analysis
  var preimageBuffer = testTx.generatePreimage(0)
  var preimageHex = preimageBuffer.toString('hex')
  
  return {
    keypair: covenantKeypair,
    utxos: utxos,
    transaction: testTx,
    preimage: {
      buffer: preimageBuffer,
      hex: preimageHex,
      length: preimageBuffer.length
    },
    
    // Covenant testing helpers
    extractField: function(fieldName) {
      try {
        var Preimage = require('./preimage')
        return Preimage.extractFromHex(preimageHex, fieldName)
      } catch (error) {
        throw new Error('SmartContract.Preimage not available: ' + error.message)
      }
    },
    
    validateCovenant: function(covenantLogic) {
      // Placeholder for covenant validation
      return {
        preimageValid: preimageBuffer.length >= 182, // Minimum BIP-143 size
        covenantLogic: covenantLogic,
        testPassed: true
      }
    },
    
    getSummary: function() {
      return {
        keypair: covenantKeypair.addressString,
        utxoCount: utxos.length,
        totalValue: utxos.reduce(function(sum, utxo) { return sum + utxo.satoshis }, 0),
        preimageLength: preimageBuffer.length,
        transactionId: testTx.transaction.id
      }
    }
  }
}

/**
 * Get all generated keypairs
 * @returns {Object} Key ring with all keypairs
 */
UTXOGenerator.prototype.getKeypairs = function() {
  return this.keyRing
}

/**
 * Get all generated UTXOs
 * @returns {Array} UTXO pool
 */
UTXOGenerator.prototype.getUTXOs = function() {
  return this.utxoPool
}

/**
 * Clear all generated data
 */
UTXOGenerator.prototype.reset = function() {
  this.keyRing = {}
  this.utxoPool = []
}

/**
 * Static utility methods
 */

/**
 * Generate a quick test setup with real BSV keys
 * @param {Object} options - Setup options
 * @returns {Object} Complete test environment
 */
UTXOGenerator.createTestEnvironment = function(options) {
  var generator = new UTXOGenerator(options)
  var covenantTest = generator.createCovenantTest(options)
  
  return {
    generator: generator,
    test: covenantTest,
    
    // Quick access methods
    getPreimage: function() {
      return covenantTest.preimage.hex
    },
    
    getKeypair: function() {
      return covenantTest.keypair
    },
    
    extractField: function(fieldName) {
      return covenantTest.extractField(fieldName)
    },
    
    generateASM: function(fieldName) {
      try {
        var Preimage = require('./preimage')
        return Preimage.generateASMFromHex(covenantTest.preimage.hex, fieldName)
      } catch (error) {
        throw new Error('Cannot generate ASM: ' + error.message)
      }
    }
  }
}

module.exports = UTXOGenerator
